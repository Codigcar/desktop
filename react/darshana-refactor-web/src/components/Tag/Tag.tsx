import { FC } from 'react';
import Image from 'next/image';

import { RenderIf } from '@utils';

import styles from './tag.module.scss';

interface IProps {
  name: string;
  background?: string;
  close?: boolean;
}

const Tag: FC<IProps> = ({ name, background, close }) => {
  return (
    <span
      className={`tag ${styles.tag}`}
      style={{ ['--bg-tag' as string]: background }}
    >
      {name}

      <RenderIf condition={close}>
        <button className={`tag--close ${styles['tag--close']}`} type="button">
          <Image
            src="/images/icons/close.svg"
            width={16}
            height={16}
            alt="Cerrar"
          />
        </button>
      </RenderIf>
    </span>
  );
};

export default Tag;
