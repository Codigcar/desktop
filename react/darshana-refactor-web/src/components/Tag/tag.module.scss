@import '@styles/settings';

.tag {
  background-color: var(--bg-tag);
  display: inline-flex;
  align-items: center;
  padding: 4px 16px;
  border-radius: 16px;

  &--close {
    display: flex;
    margin-left: 8px;
  }
}
