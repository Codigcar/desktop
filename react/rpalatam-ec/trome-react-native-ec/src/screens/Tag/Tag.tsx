import analytics from '@react-native-firebase/analytics'
import crashlytics from '@react-native-firebase/crashlytics'
import { useNavigation, useRoute } from '@react-navigation/native'
import React, { useCallback, useEffect, useMemo, useState } from 'react'

import Box from '../../components/box'
import Card, { CardPlaceholder } from '../../components/card'
import Ribbon from '../../components/ribbon'
import InfiniteScrollList, {
  ItemAdvertising,
  removeDuplicateStories,
} from '../../containers/infiniteScrollList'
import { useAuth } from '../../context/auth'
import { Advertising } from '../../entities/Advertising'
import { Story } from '../../entities/Story'
import useOpenStory from '../../hooks/useOpenStory'
import Content, { queryInclude } from '../../services/arc/Content'
import TagService from '../../services/arc/Tag'
import { getAdNewsListConfig } from '../../utils/advertising'
import { insertIntoArray } from '../../utils/tools'
import type { MainStackScreenProps } from '../../routes/types'

type Item = Advertising | Story

const { View } = Box

const TagScreen: React.FC = () => {
  const navigation = useNavigation<MainStackScreenProps<'Tag'>['navigation']>()
  const route = useRoute<MainStackScreenProps<'Tag'>['route']>()
  const { isSubscribed } = useAuth()
  const { name: tagName, path: pathname } = route.params
  const [items, setItems] = useState<Story[]>([])
  const [hasMore, setHasMore] = useState(false)

  const adSlots = useMemo(() => {
    const config = getAdNewsListConfig()
    return isSubscribed ? config.premium : config.free
  }, [isSubscribed])

  const Service = useMemo(() => {
    return Content.search({
      cache: true,
      params: {
        query: `type:story+AND+revision.published:true+AND+NOT+taxonomy.primary_section._id:(?no-flujo)+AND+NOT+taxonomy.tags.slug:"no-app"+AND+taxonomy.tags.slug:/${pathname}/`,
        sourceInclude: queryInclude('cards').join(),
      },
    })
  }, [pathname])

  useEffect(() => {
    if (tagName) return
    TagService.getBySlug(pathname).then((data) => {
      const params = { name: data?.name || 'Tag', path: pathname }
      navigation.setParams(params as Record<string, string>)
    })

    return () => {
      TagService.abort()
    }
  }, [pathname, navigation, tagName])

  const openStory = useOpenStory(tagName as string)

  const renderItem = useCallback(
    ({ item }: { item: Item }) => {
      if (item instanceof Advertising) return <ItemAdvertising item={item} />
      return (
        <Card onPress={() => openStory(item)} story={item} variant="magazine" />
      )
    },
    [openStory],
  )

  const handleLoadMore = useCallback(
    async (pageToLoad: number) => {
      try {
        const stories = await Service.get({ page: pageToLoad })
        analytics().logScreenView({
          screen_name: `screen_tag_${pathname}?page=${pageToLoad}`,
          screen_class: 'TagScreen',
        })
        setItems((prevList = []) => {
          if (pageToLoad === 0) return insertIntoArray(stories, adSlots)
          const newList = removeDuplicateStories(prevList, stories)
          return insertIntoArray(newList, adSlots)
        })
        setHasMore(stories.length >= 10)
      } catch (error) {
        if (error instanceof Error && !error.message.includes('canceled')) {
          crashlytics().recordError(error)
        }
      }
    },
    [Service, adSlots, pathname],
  )

  useEffect(() => {
    handleLoadMore(0)
    return () => {
      Service.abort()
    }
  }, [handleLoadMore, Service])

  return (
    <View bg="background" flex={1} testID="tag-screen">
      <Ribbon title={tagName} />
      <InfiniteScrollList
        data={items}
        pullRefresh
        keyExtractor={({ id }) => id}
        hasMore={hasMore && items.length < 50}
        loadMore={handleLoadMore}
        renderItem={renderItem}
        ItemSeparatorComponent={() => <View bg="separator" height={1} mx="1" />}
        ListEmptyComponent={
          <View flex={1} px="0.75">
            <CardPlaceholder repeat={8} variant="magazine" />
          </View>
        }
      />
    </View>
  )
}

export default TagScreen
