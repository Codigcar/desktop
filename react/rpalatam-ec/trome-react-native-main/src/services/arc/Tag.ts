import axios, { CancelTokenSource, CancelTokenStatic } from 'axios'
import { cacheAdapterEnhancer } from 'axios-extensions'
import LRUCache from 'lru-cache'
import NativeConfig from 'react-native-config'

export { queryInclude, getQueryBySection } from './querys'

const FIVE_MINUTES = 1000 * 60 * 5

const instance = axios.create({
  baseURL: NativeConfig.ARC_BASE_URL + '/tags/v2/slugs',
  adapter:
    axios.defaults.adapter &&
    cacheAdapterEnhancer(axios.defaults.adapter, {
      enabledByDefault: false,
      defaultCache: new LRUCache({ maxAge: FIVE_MINUTES, max: 5 }),
    }),
})

// eslint-disable-next-line dot-notation
instance.defaults.headers.common['Authorization'] = NativeConfig.ARC_TOKEN

interface TagResponse {
  description: string
  name: string
  path: string
  slug: string
  tag: string
}

class Tag {
  private cancelToken: CancelTokenStatic
  private source?: CancelTokenSource

  constructor() {
    this.cancelToken = axios.CancelToken
  }

  abort(): void {
    this.source?.cancel('Operation canceled by the user.')
  }

  async getBySlug(slug: string): Promise<TagResponse | null> {
    this.source = this.cancelToken.source()
    const response = await instance.get<{ Payload: (TagResponse | null)[] }>(
      '/',
      {
        cancelToken: this.source.token,
        params: { slugs: slug },
      },
    )
    return response.data.Payload[0]
  }
}

export default new Tag()
