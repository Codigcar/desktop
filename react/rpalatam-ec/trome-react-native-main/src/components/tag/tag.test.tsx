import { act, render } from '@testing-utils/library'
import React from 'react'

import Tag from './tag'

describe('Tag', () => {
  it('should match with her snapshot', async () => {
    const { toJSON } = render(<Tag />)

    await act(async () => {
      expect(toJSON()).toMatchSnapshot()
    })
  })
})
