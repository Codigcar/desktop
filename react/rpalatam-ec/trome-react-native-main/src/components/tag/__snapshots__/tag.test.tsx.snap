// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag should match with her snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#EF4444",
            "borderRadius": 4,
            "display": "flex",
            "marginHorizontal": 4,
            "paddingHorizontal": 8,
            "paddingVertical": 4,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "display": "flex",
              "fontFamily": "Avenir",
              "fontSize": 12,
              "fontWeight": "700",
              "lineHeight": 16,
            },
          ]
        }
      />
    </View>
  </View>
</RNCSafeAreaProvider>
`;
